### PR TITLE
Support `format_number`

### PR DESCRIPTION
**Pull request by thirtiseven**
Partly supported #9173 

This PR support format_number for integral and decimal type. Float/double support is still wip, I plan to refactor this part with a string operations based solution to avoid precision error. 

And for float/double type, the results will be mismatch between spark and plugin for large/small numbers.

It is because we should first convert a float/double to string correctly before formatting it, but in plugin casting float/double to string doesn't match Spark/Java's result, see [compatibility doc](https://github.com/NVIDIA/spark-rapids/blob/branch-23.10/docs/compatibility.md#float-to-string). We may need a custom kernel for float to string casting, see https://github.com/NVIDIA/spark-rapids/issues/4204. 

The solution is quite long and calls many times of cuDF API, which may make it slower than excepted. I did some performance test, it ran faster than CPU. 

**performance test results**

10000000 random number generated by BigDataGen:

```scala
val dataTable = DBGen().addTable("data", "a {{{type}}}", 10000000)
dataTable.toDF(spark).write.mode("overwrite").parquet("{{{type}}}_for")
```

test code:

```scala
spark.time(df.selectExpr("COUNT(format_number(a, -1)) as a", "COUNT(format_number(a, 0)) as b", "COUNT(format_number(a, 5)) as c", "COUNT(format_number(a, 50)) as d").show())
```

| Data Type | GPU Time (ms) | CPU Time (ms) |
|-|-|-|
| double | 11016 | 22905 |
| float | 1464 | 10307 |  
| int | 2743 | 5127 |
| short | 545 | 3967 |
| byte | 536 | 3481 |
| long | 3251 | 7002 |

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 1]
```diff
@@ -3086,6 +3086,19 @@ object GpuOverrides extends Logging {
          |For instance decimal strings not longer than 18 characters / hexadecimal strings
          |not longer than 15 characters disregarding the sign cannot cause an overflow.
          """.stripMargin.replaceAll("\n", " ")),
+    expr[FormatNumber](
+      "Formats numeric column x to a format like '#,###,###.##', rounded to d decimal places" +
+      " with HALF_EVEN round mode, and returns the result as a string column.\n" + 
+      "If d is 0, the result has no decimal point or fractional part. If d is less than 0, " + 
+      "the result will be null.",
+      ExprChecks.binaryProject(TypeSig.STRING, TypeSig.STRING,
+        ("x", TypeSig.cpuNumeric, TypeSig.cpuNumeric),
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala]: Can we change this to gpuNumeric for the GPU version? They are the same so it should not change anything, but to be defensive it would be good to keep them separate in case spark adds a new number format.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 2]
```diff
@@ -3086,6 +3086,19 @@ object GpuOverrides extends Logging {
...
+        ("d", TypeSig.lit(TypeEnum.INT), TypeSig.INT)),
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3093]: For the CPU side we should include TypeSig.STRING too so it is clear that we have only partial support and in what ways.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 3]
```diff
@@ -2079,4 +2078,601 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+  
+  override def left: Expression = x
+  override def right: Expression = d
+  override def dataType: DataType = StringType
+  override def nullable: Boolean = true
+  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, IntegerType)
+
+  private def stripLeadingZeros(cv: ColumnVector): ColumnVector = {
+    val (isZero, zeroCv) = withResource(Scalar.fromString("0")) { zero =>
+      (cv.startsWith(zero), ColumnVector.fromScalar(zero, cv.getRowCount.toInt))
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2095]: if `fromScalar` fails, like with an OOM the result of startsWith will be leaked.  To make this work we would need to rewrite it as.

```
  val (isZero, zeroCv) = withResource(Scalar.fromString("0")) { zero =>
    closeOnExcept(cv.startsWith(zero)) { isZero =>
      (isZero, ColumnVector.fromScalar(zero, cv.getRowCount.toInt))
    }
  }
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 4]
```diff
@@ -2079,4 +2078,601 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+case class GpuFormatNumber(x: Expression, d: Expression)
+    extends GpuBinaryExpression with ExpectsInputTypes with NullIntolerant {
+  
+  override def left: Expression = x
+  override def right: Expression = d
+  override def dataType: DataType = StringType
+  override def nullable: Boolean = true
+  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, IntegerType)
+
+  private def stripLeadingZeros(cv: ColumnVector): ColumnVector = {
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: Could you add some comments here, because the name of the function appears to be off from what it actually does. Converting 0000 to 0 is fine, but 00010 to 0 is not what I would expect from stripLeadingZeros.

File: docs/compatibility.md [hunk 5]
```diff
@@ -661,6 +661,10 @@ The GPU will use different precision than Java's toString method when converting
 types to strings. The GPU uses a lowercase `e` prefix for an exponent while Spark uses uppercase
 `E`. As a result the computed string can differ from the default behavior in Spark.
 
+The `format_number` function will retain 10 digits of precision for the GPU when the input is a floating
```
Comment (revans2) [inline docs/compatibility.md:664]: Can we have floating point be off by default then, so users can opt into it?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 6]
```diff
@@ -2079,4 +2078,620 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+    extends GpuBinaryExpression with ExpectsInputTypes with NullIntolerant {
+  
+  override def left: Expression = x
+  override def right: Expression = d
+  override def dataType: DataType = StringType
+  override def nullable: Boolean = true
+  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, IntegerType)
+
+  private def reduceLeadingZeros(cv: ColumnVector): ColumnVector = {
+    // replace "[0]+" to "0" in a column vector
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2093]: But it does not really do that, unless you are guaranteed that the input never starts with a 0. Could we have better docs about this?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 7]
```diff
@@ -2079,4 +2078,620 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+    // get three parts first:
+    val replaceDelimToE = withResource(Scalar.fromString("e")) { e =>
+      withResource(Scalar.fromString(".")) { p =>
+        cv.stringReplace(e, p)
+      }
+    }
+    // get three parts: 1.234567e+7 -> 1, 234567, +7
+    val (intPartSign, decPart, expPart) = withResource(replaceDelimToE) { _ =>
+      withResource(replaceDelimToE.stringSplit(".", 3)) { intDecExp =>
+        (intDecExp.getColumn(0).incRefCount(),
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2139]: I don't think this works. I thought stringSplit returned an array, not a struct, so there is just one child column of an array.

**Review by revans2 — COMMENTED** [review]
I only got part way through. I'll finish the review when it is no longer in draft.

**Comment by sameerz** [comment]
> I did some performance test on long type, it ran faster than CPU. I will do more tests and update soon.

Can you please add the performance test results here?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala (hunk 7 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2139]: I think `stringSplit` returned a `Table` so we can getColumn from it. [cuDF doc](https://docs.rapids.ai/api/cudf-java/stable/ai/rapids/cudf/columnview#stringSplit-java.lang.String-int-)

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala (hunk 6 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: Added a comment and moved this back to `getPartsFromDecimal` because it is ad-hoc and only used once.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala (hunk 3 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: thanks, done.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala (hunk 1 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala]: done

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 8]
```diff
@@ -2079,4 +2078,621 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+        withResource(intPartNegExp) { _ =>
+          expPos.ifElse(intPartPosExp, intPartNegExp)
+        }
+      }
+      val decPartExp = withResource(decPartPosExp) { _ =>
+        withResource(decPartNegExp) { _ =>
+          expPos.ifElse(decPartPosExp, decPartNegExp)
+        }
+      }
+      (intPartExp, decPartExp)
```
Comment (res-life) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2374]: If execption occurs at line 2371, then `intPartExp` will be leaked.
Should handle this resource pair as a whole.
```
      withResource(ArrayBuffer.empty[AutoCloseable]) { resource_array =>
        
        var res1 = genRes()
        resource_array ++= res1
        
        var res2 = genRes()
        resource_array ++= res2
        
      }
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 9]
```diff
@@ -2079,4 +2078,621 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+            withResource(normalInt) { _ =>
+              containsE.ifElse(expInt, normalInt)
+            }
+          }
+          val decPart = withResource(expDec) { _ =>
+            withResource(normalDec) { _ =>
+              containsE.ifElse(expDec, normalDec)
+            }
+          }
+          (intPart, decPart)
```
Comment (res-life) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2424]: Handle resource pair as a whole.

**Review by revans2 — COMMENTED** [review]
I started to go through this, but the complexity, especially for the float/double code is rather hard to follow.  Also in my own testing I saw a lot of problems with the float/double results. It looks like it is related to rounding and to the amount of precision we can get as output from casting a float to a string using CUDF. I'm not sure if we can fix that without a custom kernel.

**Review by firestarman — COMMENTED** [review]
More comment for the processing will be better for review.

**Comment by thirtiseven** [comment]
> > I did some performance test on long type, it ran faster than CPU. I will do more tests and update soon.
> 
> Can you please add the performance test results here?

Ok, updated them in the PR description.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 10]
```diff
@@ -2079,4 +2078,654 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+      withResource(strlen.max()) { maxlen =>
+        maxlen.isValid match {
+          case true => maxlen.getInt
+          case false => 0
+        }
+      }
+    }
+    val substrs = (0 until maxstrlen by 3).map { i =>
+      val endIndex = i + 3
+      str.substring(i, endIndex).asInstanceOf[ColumnView]
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: It is because array is invariant in scala. I will got:
```
[ERROR] [Error] /home/haoyangl/spark-rapids/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2583: type mismatch;
 found   : Array[ai.rapids.cudf.ColumnVector]
 required: Array[ai.rapids.cudf.ColumnView]
Note: ai.rapids.cudf.ColumnVector <: ai.rapids.cudf.ColumnView, but class Array is invariant in type T.
You may wish to investigate a wildcard type such as `_ <: ai.rapids.cudf.ColumnView`. (SLS 3.2.10)
```
if removing this.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 11]
```diff
@@ -2079,4 +2078,661 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+          case true => maxlen.getInt
+          case false => 0
+        }
+      }
+    }
+    val sepCol = withResource(Scalar.fromString(",")) { sep =>
+      ColumnVector.fromScalar(sep, str.getRowCount.toInt)
+    }
+    val substrs = closeOnExcept(sepCol) { _ =>
+      (0 until maxstrlen by 3).map { i =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2577]: I think we do not need to reverse and reverse back the input string if we slice string from end to start, something like
```
var curEndsCol: Option[ColumnVector] = strlen

withResource(curEndsCol) { _ =>
  (0 until maxstrlen by 3).safeMap { _ =>
    val startCol = curEndsCol - 3,
    val sub = closeOnExcept(startCol) { _ =>
      str.substring(startCol, curEndsCol)
    }
    curEndsCol.close()
    curEndsCol = startCol
    sub
  }
}
```
You need to do `strlen - 3` in columnar way, e.g.
```
withResource(Scalar.fromInt(3)) { scalar3 =>
   strlen.sub(scalar3)
}
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 12]
```diff
@@ -2079,4 +2078,654 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+    val substrs = (0 until maxstrlen by 3).map { i =>
+      val endIndex = i + 3
+      str.substring(i, endIndex).asInstanceOf[ColumnView]
+    }.toArray
+    val sepCol = withResource(Scalar.fromString(",")) { sep =>
+      ColumnVector.fromScalar(sep, str.getRowCount.toInt)
+    }
+    withResource(substrs) { _ =>
+      withResource(sepCol) { _ =>
+        withResource(ColumnVector.stringConcatenate(substrs, sepCol)) { res =>
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2583]: I think in that overload case `substrs` is required to have two or more columns, but it is often have only one column here. [cuDF code](https://github.com/rapidsai/cudf/blob/31e56702fe15f44b3e849207d31d0bb79c307367/cpp/src/strings/combine/concatenate.cu#L128)

[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2583]: OK, but it is a quite strange requirement.

File: docs/compatibility.md (hunk 5 above)
Comment (thirtiseven) [reply docs/compatibility.md:664]: done

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 13]
```diff
@@ -2079,4 +2078,654 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+    handleNegInf
+  }
+
+  override def doColumnar(lhs: GpuColumnVector, rhs: GpuScalar): ColumnVector = {
+    // get int d from rhs
+    if (!rhs.isValid || rhs.getValue.asInstanceOf[Int] < 0) {
+      return GpuColumnVector.columnVectorFromNull(lhs.getRowCount.toInt, StringType)
+    }
+    val d = rhs.getValue.asInstanceOf[Int]
+    val isNegative = negativeCheck(lhs.getBase)
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: done.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 14]
```diff
@@ -2079,4 +2078,660 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+          reverseBack
+        }
+        case _ => {
+          // d > 0, append decimal part to result
+          val pointCv = withResource(Scalar.fromString(".")) { point =>
+            ColumnVector.fromScalar(point, lhs.getRowCount.toInt)
+          }
+          withResource(reverseBack) { _ =>
+            withResource(pointCv) { _ =>
+              ColumnVector.stringConcatenate(Array(reverseBack, pointCv, decimalPart))
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: done.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 15]
```diff
@@ -2079,4 +2078,654 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+      case true => {
+        withResource(formated) { _ =>
+          withResource(lhs.getBase.isNull) { isNull =>
+            withResource(Scalar.fromNull(DType.STRING)) { nullScalar =>
+              isNull.ifElse(nullScalar, formated)
+            }
+          }
+        }
+      }
+      case false => {
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: done.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 16]
```diff
@@ -2079,4 +2078,654 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+      }
+    }
+    // handle inf and nan
+    x.dataType match {
+      case FloatType | DoubleType => {
+        withResource(formatedWithNull) { _ =>
+          handleInfAndNan(lhs.getBase, formatedWithNull)
+        }
+      }
+      case _ => {
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: done.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 17]
```diff
@@ -2079,4 +2078,654 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+  private def addCommas(str: ColumnVector): ColumnVector = {
+    val maxstrlen = withResource(str.getCharLengths()) { strlen =>
+      withResource(strlen.max()) { maxlen =>
+        maxlen.isValid match {
+          case true => maxlen.getInt
+          case false => 0
+        }
+      }
+    }
+    val substrs = (0 until maxstrlen by 3).map { i =>
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: done.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 18]
```diff
@@ -2079,4 +2078,654 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+        ColumnVector.stringConcatenate(Array(negCv, resWithDecimalPart))
+      }
+      withResource(isNegative) { _ =>
+        withResource(resWithNeg) { _ =>
+          isNegative.ifElse(resWithNeg, resWithDecimalPart)
+        }
+      }
+    }
+    // handle null case
+    val formatedWithNull = nullable match {
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: done.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 19]
```diff
@@ -2079,4 +2078,659 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+    }
+    val sepCol = withResource(Scalar.fromString(",")) { sep =>
+      ColumnVector.fromScalar(sep, str.getRowCount.toInt)
+    }
+    val substrs = closeOnExcept(sepCol) { _ =>
+      (0 until maxstrlen by 3).safeMap { i =>
+        str.substring(i, i + 3).asInstanceOf[ColumnView]
+      }.toArray
+    }
+    // TODO: test Liangcai's solution here
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2578]: Lets not leave commented out code in here. If you want to test it either do it now, or file a follow on issue to do it.

**Review by revans2 — COMMENTED** [review]
I think this is very close.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 20]
```diff
@@ -2079,4 +2078,654 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+      }
+      case DecimalType.Fixed(_, scale) => {
+        getPartsFromDecimal(cv, d, scale)
+      }
+      case IntegerType | LongType | ShortType | ByteType => {
+        closeOnExcept(ArrayBuffer.empty[ColumnVector]) { resource_array =>
+          val intPartPos = withResource(cv.castTo(DType.STRING)) { intPart =>
+            withResource(Scalar.fromString("-")) { negativeSign =>
+              withResource(Scalar.fromString("")) { emptyString =>
+                intPart.stringReplace(negativeSign, emptyString)
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: Because sign will be handled later for all types, mainly because the float -0.0/0.0 case. I'm reconsidering to handle the sign inside `getParts`.

I updated here to `lstrip` now, looks better.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 21]
```diff
@@ -2079,4 +2078,667 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+          withResource(getZeroCv(cv.getRowCount.toInt)) { zeroCv =>
+            zeroCv.repeatStrings(zerosNumRounding)
+          }
+        }
+        val zeroPointCv = withResource(Scalar.fromString("0.")) { point =>
+          ColumnVector.fromScalar(point, cv.getRowCount.toInt)
+        }
+        val numberToRoundStr = withResource(zeroPointCv) { _ =>
+          withResource(leadingZeros) { _ =>
+            ColumnVector.stringConcatenate(Array(zeroPointCv, leadingZeros, intPart, decPart))
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2263]: nit: scalar version is better.

File: integration_tests/src/main/python/string_test.py [hunk 22]
```diff
@@ -797,3 +797,42 @@ def test_conv_dec_to_from_hex(from_base, to_base, pattern):
...
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark: unary_op_df(spark, gen).selectExpr(
+            'format_number(a, 5)'),
+        conf = float_format_number_conf
+    )
+
+@allow_non_gpu('ProjectExec')
+def test_format_number_float_fallback():
+    gen = DoubleGen()
+    assert_gpu_fallback_collect(
```
Comment (firestarman) [inline integration_tests/src/main/python/string_test.py:834]: Why will this case fall back ?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 23]
```diff
@@ -3086,6 +3086,29 @@ object GpuOverrides extends Logging {
          |For instance decimal strings not longer than 18 characters / hexadecimal strings
          |not longer than 15 characters disregarding the sign cannot cause an overflow.
          """.stripMargin.replaceAll("\n", " ")),
+    expr[FormatNumber](
+      "Formats the number x like '#,###,###.##', rounded to d decimal places.",
+      ExprChecks.binaryProject(TypeSig.STRING, TypeSig.STRING,
+        ("x", TypeSig.gpuNumeric, TypeSig.cpuNumeric),
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3092]: We may need to disable DecimalType support temporarily due to the bug https://github.com/NVIDIA/spark-rapids/issues/9309

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 24]
```diff
@@ -2079,4 +2078,659 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+              }
+            }
+            resource_array.clear()
+            resource_array += intPart
+            val decPart = withResource(expDec) { _ =>
+              withResource(normalDec) { _ =>
+                containsE.ifElse(expDec, normalDec)
+              }
+            }
+            (intPart, decPart)
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2451]: It is in a closeOnExcept too.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 25]
```diff
@@ -2079,4 +2078,659 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
...
+        // if no special case, return normal case directly
+        val anyExp = closeOnExcept(cvStr) { _ =>
+          withResource(containsE.any()) { any =>
+            any.isValid && any.getBoolean
+          }
+        }
+        anyExp match {
+          case false => {
+            cvStr.safeClose()
+            (normalInt, normalDec)
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2421]: I think they are in a closeOnExcept so no need to increment reference count. (Not very sure whether the usage `closeOnExcept(ArrayBuffer.empty[ColumnVector])` is correct.)

**Comment by thirtiseven** [comment]
@revans2  Thanks for the review, I think I fixed the memory issues.

> I'm not sure we can fix this without a custom kernel.

I don't think we can match Spark's result on the plugin side for now. This PR can produce correct results with limited precision and aim to require least change to fully support double/float when float to string in JNI is ready.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala (hunk 23 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3092]: If you agree, I'd like to update it in a follow-on pr (maybe disable it along with `round`) if this pr is going to be merged.

Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3092]: ok, but usually need an issue to track this followup.

Comment (thirtiseven) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3092]: filed #9322

Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3092]: I would prefer to add it in this PR, since we already know there is an issue for DecimalType.

Comment (thirtiseven) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3092]: updated.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 26]
```diff
@@ -3086,6 +3086,35 @@ object GpuOverrides extends Logging {
...
+                willNotWorkOnGpu("format_number with floating point types on the GPU returns " +
+                  "results that have a different precision than the default results of Spark. " +
+                  "To enable this operation on the GPU, set" +
+                  s" ${RapidsConf.ENABLE_FLOAT_FORMAT_NUMBER} to true.")
+              }
+            }
+            case dt: DecimalType => {
+              if (dt.scale > 32) {
+                willNotWorkOnGpu("format_number will generate results mismatched from Spark " +
+                  "when the scale is large.")
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3108]: We know the exact scale value,  so we can print it in the message string.
```suggestion
                  "when the scale is larger than 32.")
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 27]
```diff
@@ -3102,6 +3102,12 @@ object GpuOverrides extends Logging {
                   s" ${RapidsConf.ENABLE_FLOAT_FORMAT_NUMBER} to true.")
               }
             }
+            case dt: DecimalType => {
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3105]: Once https://github.com/rapidsai/cudf/issues/14210 is fixed we should come back and retest to be sure that it is working properly.

**Review by revans2 — APPROVED** [review]
Looks good to me.

**Comment by thirtiseven** [comment]
Thanks all for the review and help! merging this...
